src/relations/database_provides.py

```diff
@@ -0,0 +1,151 @@
+# Copyright 2022 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Library containing the implementation of the database provides relation."""
+
+import json
+import logging
+
+from charms.data_platform_libs.v0.database_provides import (
+    DatabaseProvides,
+    DatabaseRequestedEvent,
+)
+from ops.framework import Object
+from ops.model import Application, BlockedStatus
+
+from constants import (
+    DATABASE_PROVIDES_RELATION,
+    MYSQL_ROUTER_LEADER_BOOTSTRAPED,
+    MYSQL_ROUTER_PROVIDES_DATA,
+    MYSQL_ROUTER_REQUIRES_DATA,
+    PASSWORD_LENGTH,
+    PEER,
+)
+from mysql_router_helpers import (
+    MySQLRouter,
+    MySQLRouterBootstrapError,
+    MySQLRouterCreateUserWithDatabasePrivilegesError,
+)
+from utils import generate_random_password
+
+logger = logging.getLogger(__name__)
+
+
+class DatabaseProvidesRelation(Object):
+    """Encapsulation of the relation between mysqlrouter and the consumer application."""
+
+    def __init__(self, charm):
+        super().__init__(charm, DATABASE_PROVIDES_RELATION)
+
+        self.charm = charm
+        self.database = DatabaseProvides(self.charm, relation_name=DATABASE_PROVIDES_RELATION)
+
+        self.framework.observe(self.database.on.database_requested, self._on_database_requested)
+
+        self.framework.observe(
+            self.charm.on[PEER].relation_changed, self._on_peer_relation_changed
+        )
+
+    # =======================
+    #  Helpers
+    # =======================
+
+    def _database_provides_relation_exists(self) -> bool:
+        database_provides_relations = self.charm.model.relations.get(DATABASE_PROVIDES_RELATION)
```

Comment:
right, that's an assumption i made (since it's a subordinate charm, there's only either one related app or none). each application (if there are multiple on the same virtual machine) should have their own mysqlrouter applications